File: src/scpn_theta_pinch_core/physics/balance.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SCPN Theta-Pinch Core — sharp-boundary state

"""Sharp-boundary operating state of a theta-pinch column.

The declared plasma pressure and compression field of the validated
configuration fix the sharp-boundary beta ``beta = p / (B^2 / 2 mu0)``;
with a declared ion density and ion mass they fix the equal-species ion
temperature ``T = p / (2 n e)``, the Alfvén speed
``v_A = B / sqrt(mu0 n m_i)`` and the end-to-centre Alfvén propagation
time ``tau_A = (L / 2) / v_A`` that the Scyllac review uses to order the
onset of end effects (W. E. Quinn et al., LA-UR-73-1053 (1973), p. 14:
"the time for propagation of an Alfvén wave from the ends of the coil").
Every sharp-boundary relation of this package needs ``0 < beta < 1``;
``beta = 1`` (allowed by the configuration invariant) is refused here,
never clamped.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Any, Final

from scpn_theta_pinch_core.configuration import DeviceConfiguration
from scpn_theta_pinch_core.errors import DeviceConfigurationError
from scpn_theta_pinch_core.parameters import MU0, require_positive

ELEMENTARY_CHARGE_C: Final = 1.602176634e-19
PROTON_MASS_KG: Final = 1.67262192369e-27
DEUTERON_MASS_KG: Final = 3.3435837724e-27


def require_sharp_boundary_beta(beta: float) -> float:
    """Return ``beta`` when strictly inside the sharp-boundary domain.

    Parameters
    ----------
    beta
        Sharp-boundary beta of the operating point.

    Returns
    -------
    float
        The validated beta.

    Raises
    ------
    DeviceConfigurationError
        If ``beta`` is not strictly between zero and one; the relations
        of Quinn et al. (1973) carry ``1 - beta`` denominators.
    """
    if not 0.0 < beta < 1.0:
        raise DeviceConfigurationError(
            f"beta: sharp-boundary relations require 0 < beta < 1, got {beta!r}"
        )
    return beta


@dataclass(frozen=True, slots=True)
class SharpBoundaryState:
    """Operating state derived from the configuration and declared inputs.

    Parameters
    ----------
    beta
        ``p / (B^2 / 2 mu0)``.
    field_t
        Compression field ``B`` in tesla.
    plasma_pressure_pa
        Declared plasma pressure.
    ion_density_per_m3
        Declared ion density ``n``.
    ion_temperature_ev
        ``T = p / (2 n e)`` for equal species.
    alfven_speed_m_s
        ``B / sqrt(mu0 n m_i)``.
    end_alfven_time_s
        ``(L / 2) / v_A`` for the coil length ``L``.
    """

    beta: float
    field_t: float
    plasma_pressure_pa: float
    ion_density_per_m3: float
    ion_temperature_ev: float
    alfven_speed_m_s: float
    end_alfven_time_s: float

    def to_record(self) -> dict[str, Any]:
        """Project the state to a JSON-serialisable record.

        Returns
        -------
        dict[str, Any]
            Every field under its name.
        """
        return {
            "beta": self.beta,
            "field_t": self.field_t,
            "plasma_pressure_pa": self.plasma_pressure_pa,
            "ion_density_per_m3": self.ion_density_per_m3,
            "ion_temperature_ev": self.ion_temperature_ev,
            "alfven_speed_m_s": self.alfven_speed_m_s,
            "end_alfven_time_s": self.end_alfven_time_s,
        }


def sharp_boundary_state(
    configuration: DeviceConfiguration, ion_mass_kg: float, ion_density_per_m3: float
) -> SharpBoundaryState:
    """Evaluate the sharp-boundary state of a validated configuration.

    Parameters
    ----------
    configuration
        Validated theta-pinch configuration.
    ion_mass_kg
        Ion mass; strictly positive.
    ion_density_per_m3
        Declared ion density; strictly positive.

    Returns
    -------
    SharpBoundaryState
        Beta, ion temperature, Alfvén speed and end propagation time.

    Raises
    ------
    DeviceConfigurationError
        If an input is invalid or beta lies outside ``(0, 1)``.
    """
    require_positive("ion_mass_kg", ion_mass_kg)
    require_positive("ion_density_per_m3", ion_density_per_m3)
    beta = require_sharp_boundary_beta(configuration.beta())
    field_t = configuration.coil.coil_field_t
    pressure = configuration.plasma.plasma_pressure_pa
    ion_temperature_ev = pressure / (2.0 * ion_density_per_m3 * ELEMENTARY_CHARGE_C)
    alfven_speed_m_s = field_t / math.sqrt(MU0 * ion_density_per_m3 * ion_mass_kg)
    end_alfven_time_s = (configuration.coil.coil_length_m / 2.0) / alfven_speed_m_s
    return SharpBoundaryState(
        beta=beta,
        field_t=field_t,
        plasma_pressure_pa=pressure,
        ion_density_per_m3=ion_density_per_m3,
        ion_temperature_ev=ion_temperature_ev,
        alfven_speed_m_s=alfven_speed_m_s,
        end_alfven_time_s=end_alfven_time_s,
    )

File: rust/src/balance.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SCPN Theta-Pinch Core — sharp-boundary state kernel

//! Sharp-boundary state (Quinn et al. 1973, p. 14 for the Alfvén end
//! propagation time), operation-for-operation identical to
//! `scpn_theta_pinch_core.physics.balance.sharp_boundary_state`.

use crate::{ELEMENTARY_CHARGE_C, MU0};

/// Sharp-boundary operating state (SI units in the names).
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct SharpBoundaryState {
    /// `p / (B^2 / 2 mu0)`.
    pub beta: f64,
    /// `p / (2 n e)` in electronvolts.
    pub ion_temperature_ev: f64,
    /// `B / sqrt(mu0 n m_i)`.
    pub alfven_speed_m_s: f64,
    /// `(L / 2) / v_A`.
    pub end_alfven_time_s: f64,
}

/// Evaluate the sharp-boundary state.
///
/// Inputs are validated by the Python floor (finite, strictly positive,
/// `0 < beta < 1`); the kernel assumes them.
#[must_use]
pub fn sharp_boundary_state(
    coil_field_t: f64,
    coil_length_m: f64,
    plasma_pressure_pa: f64,
    ion_mass_kg: f64,
    ion_density_per_m3: f64,
) -> SharpBoundaryState {
    let magnetic_pressure_pa = coil_field_t * coil_field_t / (2.0 * MU0);
    let beta = plasma_pressure_pa / magnetic_pressure_pa;
    let ion_temperature_ev = plasma_pressure_pa / (2.0 * ion_density_per_m3 * ELEMENTARY_CHARGE_C);
    let alfven_speed_m_s = coil_field_t / (MU0 * ion_density_per_m3 * ion_mass_kg).sqrt();
    let end_alfven_time_s = (coil_length_m / 2.0) / alfven_speed_m_s;
    SharpBoundaryState {
        beta,
        ion_temperature_ev,
        alfven_speed_m_s,
        end_alfven_time_s,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn beta_and_temperature_close() {
        let b = 3.6;
        let n = 2.5e22;
        let p = 0.85 * b * b / (2.0 * MU0);
        let s = sharp_boundary_state(b, 5.0, p, 3.343_583_772_4e-27, n);
        assert!((s.beta - 0.85).abs() <= 1.0e-15);
        let t_ev = p / (2.0 * n * ELEMENTARY_CHARGE_C);
        assert_eq!(s.ion_temperature_ev, t_ev);
        assert!(s.alfven_speed_m_s > 3.0e5 && s.alfven_speed_m_s < 4.0e5);
        assert!((s.end_alfven_time_s - 2.5 / s.alfven_speed_m_s).abs() <= 1.0e-20);
    }
}
